serialized_data_interface/relation.py

```diff
@@ -0,0 +1,297 @@
+        return list(self.charm.model.relations[self.endpoint])
+
+    @cache
+    def get_status(self, relation: Relation):
+        """Get the suggested status for the given Relation."""
+        if self.is_failed(relation):
+            return BlockedStatus(f"Error handling relation: {relation.name}")
+        elif not self.is_available(relation):
+            if relation.units:
+                # If we have remote units but still no version, then there's
+                # probably something wrong and we should be blocked.
+                return BlockedStatus(f"Missing relation versions: {relation.name}")
+            else:
+                # Otherwise, we might just not have seen the versions yet.
+                return WaitingStatus(f"Waiting on relation: {relation.name}")
+        elif not self.is_ready(relation):
+            return WaitingStatus(f"Waiting on relation: {relation.name}")
+        return ActiveStatus()
+
+    @cache
+    def is_available(self, relation: Relation = None):
+        """Checks whether the given relation, or any relation if not specified, is available.
+
+        A given relation is available if the version negotation has succeeded.
+        """
+        if relation is None:
+            return any(self.is_available(relation) for relation in self.relations)
+        try:
+            self._sdi.get_version(relation)
+        except errors.RelationException:
+            return False
+        else:
+            return True
+
+    @cache
+    def is_ready(self, relation: Relation = None):
+        """Checks whether the given relation, or any relation if not specified, is ready.
+
+        A given relation is ready if the remote side has sent valid data.
+        """
+        if relation is None:
+            return any(self.is_ready(relation) for relation in self.relations)
+        try:
+            data = self.unwrap(relation)
+        except errors.RelationException:
+            return False
+        else:
+            return any(
+                data[entity] for entity in data if entity not in (self.app, self.unit)
+            )
+
+    @cache
+    def is_failed(self, relation: Relation = None):
+        """Checks whether the given relation, or any relation if not specified, has an error."""
+        if relation is None:
+            return any(self.is_failed(relation) for relation in self.relations)
+        try:
+            self.unwrap(relation)
+        except errors.RelationError as e:
+            log.exception(f"Error handling relation: {e}")
+            return True
+        except errors.UnversionedRelation:
+            if relation.units:
+                # If we have remote units but still no version, then there's
+                # probably something wrong.
+                return True
+            else:
+                return False
+        except errors.RelationException:
+            return False
+        else:
+            return False
```

Comment:
i was just surprised that anything qualifying as a `RelationException` here meant that `is_failed(..) == False` ... implying it wasn't a failure.

Reply:
I ended up removing the `IncompleteRelation` exception, since it wasn't being used. I could see an argument for removing `UnversionedRelation` as well, since yes, as you note it's part of the normal back-and-forth flow of setting up the relation and therefore not exceptional. However, it *is* currently used to distinguish between the case of the relation not being at the point where data can be exchanged vs being at that point but just not having any data yet. Plus it's required for backward compatibility, while `IncompleteRelation` was not.